src/webgpu/api/validation/vertex_state.spec.ts

```diff
@@ -488,183 +459,202 @@ g.test('vertex_shader_input_location_in_vertex_state')
+        const halfAlignment = Math.floor(bytesPerComponent / 2);
+        if (halfAlignment !== 0) {
+          offsetsToTest.push(halfAlignment);
+          offsetsToTest.push(p.arrayStride - formatSize - halfAlignment);
+        }
+
+        return poptions('offset', offsetsToTest);
+      })
+  )
+  .subcases(() =>
+    params()
+      .combine(poptions('vertexBufferIndex', [0, 1, kMaxVertexBuffers - 1]))
+      .combine(poptions('extraAttributeCount', [0, 1, kMaxVertexAttributes - 1]))
+      .combine(pbool('testAttributeAtStart'))
+  )
+  .fn(t => {
+    const {
+      format,
+      arrayStride,
+      offset,
+      vertexBufferIndex,
+      extraAttributeCount,
+      testAttributeAtStart,
+    } = t.params;
 
-g.test('check_multiple_of_4_bytes_constraint_on_offset').fn(async t => {
-  const vertexState = {
-    vertexBuffers: [
-      {
-        arrayStride: 0,
-        attributes: [
-          { format: 'float32' as GPUVertexFormat, offset: SIZEOF_FLOAT, shaderLocation: 0 },
-        ],
-      },
-    ],
-  };
-  {
-    // Control case, setting offset 4 bytes
-    const descriptor = t.getDescriptor(vertexState, VERTEX_SHADER_CODE_WITH_NO_INPUT);
-    t.device.createRenderPipeline(descriptor);
-  }
-  {
-    // Test offset of 2 bytes with uint8x2 format
-    vertexState.vertexBuffers[0].attributes[0].offset = 2;
-    vertexState.vertexBuffers[0].attributes[0].format = 'uint8x2';
-    const descriptor = t.getDescriptor(vertexState, VERTEX_SHADER_CODE_WITH_NO_INPUT);
-    t.expectValidationError(() => {
-      t.device.createRenderPipeline(descriptor);
+    const attributes: GPUVertexAttributeDescriptor[] = [];
+    addTestAttributes(attributes, {
+      testAttribute: { format, offset, shaderLocation: 0 },
+      testAttributeAtStart,
+      extraAttributeCount,
+      extraAttributeSkippedLocations: [0],
     });
-  }
-  {
-    // Test offset of 2 bytes with float32 format
-    vertexState.vertexBuffers[0].attributes[0].offset = 2;
-    vertexState.vertexBuffers[0].attributes[0].format = 'float32';
-    const descriptor = t.getDescriptor(vertexState, VERTEX_SHADER_CODE_WITH_NO_INPUT);
 
-    t.expectValidationError(() => {
-      t.device.createRenderPipeline(descriptor);
+    const vertexBuffers = [];
+    vertexBuffers[vertexBufferIndex] = { arrayStride, attributes };
+
+    const success = offset % kVertexFormatInfo[format].bytesPerComponent === 0;
+    t.testVertexState(success, { vertexBuffers });
+  });
+
+g.test('vertex_attribute_contained_in_stride')
+  .desc(
+    `
+    Test that vertex attribute [offset, offset + formatSize) must be contained in the arrayStride if arrayStride is not 0:
+    - Test for all formats.
+    - Test for various arrayStrides and offsets within that stride
+    - Test for various vertex buffer indices
+    - Test for various amounts of attributes in that vertex buffer`
+  )
+  .cases(
+    params()
+      .combine(poptions('format', kVertexFormats))
+      .combine(
```

Comment:
nit: Probably arrayStride and offset (at least offset) make more sense as subcases

Reply:
It's a little weird but it can be done, see the commit I pushed

Reply:
(I have ideas about how to improve this - subcases could be passed a pre-constructed single-value params() object so you can just `p => p.combine`, but haven't had a chance to try this)

Reply:
Thanks for the fix!